File: Java/242. Valid Anagram.java
/**
 * https://leetcode.com/problems/valid-anagram/
 */

class Solution {
    // O(n) time complexity
    // O(1) space complexity (O(26) space for the array)
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;
        int[] arr = new int[26];
        for (int i = 0; i < s.length(); i++) {
            arr[s.charAt(i) - 'a']++;
            arr[t.charAt(i) - 'a']--;
        }
        for (int a : arr)
            if (a != 0) return false;
        return true;
    }
}

File: Kotlin/242. Valid Anagram.kt
/**
 * https://leetcode.com/problems/valid-anagram/
 */

class Solution {
    // O(n) time complexity
    // O(1) space complexity
    fun isAnagram(s: String, t: String): Boolean {
        if (s.length != t.length) return false
        val arr = IntArray(26)
        for (i in s.indices) {
            arr[s[i] - 'a']++
            arr[t[i] - 'a']--
        }
        for (i in arr.indices)
            if (arr[i] != 0) return false
        return true
    }
}